drivers/DeviceKey.cpp

```diff
@@ -0,0 +1,255 @@
+int DeviceKey::calculate_cmac(const unsigned char *input, size_t isize, uint32_t *ikey_buff, int ikey_size,
+                              unsigned char *output)
+{
+    int ret;
+    mbedtls_cipher_context_t ctx;
+
+    mbedtls_cipher_type_t mbedtls_cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type = MBEDTLS_CIPHER_AES_256_ECB;
+    }
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(mbedtls_cipher_type);
+
+    mbedtls_cipher_init(&ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_starts(&ctx, (unsigned char *)ikey_buff, ikey_size * 8);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_update(&ctx, input, isize);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_finish(&ctx, output);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    return DEVICEKEY_SUCCESS;
+
+finish:
+    mbedtls_cipher_free( &ctx );
+    return ret;
+}
+
+int DeviceKey::get_derived_key(uint32_t *ikey_buff, size_t ikey_size, const unsigned char *isalt,
+                               size_t isalt_size, unsigned char *output, uint32_t ikey_type)
+{
+    int ret;
+    unsigned char *double_size_salt = NULL;
+
+    if (DEVICE_KEY_16BYTE == ikey_type) {
+        ret = calculate_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
+        if (DEVICEKEY_SUCCESS != ret) {
+            goto finish;
+        }
+    }
+
+    if (DEVICE_KEY_32BYTE == ikey_type) {
+        ret = this->calculate_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
+        if (DEVICEKEY_SUCCESS != ret) {
+            goto finish;
+        }
+
+        //Double the salt size cause cmac always return just 16 bytes
+        double_size_salt = new unsigned char[isalt_size * 2];
+        memcpy(double_size_salt, isalt, isalt_size);
+        memcpy(double_size_salt + isalt_size, isalt, isalt_size);
+
+        ret = this->calculate_cmac(double_size_salt, isalt_size * 2, ikey_buff, ikey_size, output + 16);
+    }
+
+finish:
+    if (double_size_salt != NULL) {
+        delete[] double_size_salt;
+    }
+
+    if (DEVICEKEY_SUCCESS != ret) {
+        return DEVICEKEY_ERR_CMAC_GENERIC_FAILURE;
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::generate_key_by_trng(uint32_t *output, size_t& size)
+{
+#if defined(DEVICE_TRNG)
+    size_t in_size;
+    trng_t trng_obj;
+
+    memset(output, 0, size);
+
+    if (DEVICE_KEY_16BYTE > size) {
+        return DEVICEKEY_BUFFER_TO_SMALL;
+    } else if (DEVICE_KEY_16BYTE <= size && DEVICE_KEY_32BYTE > size) {
+        in_size = DEVICE_KEY_16BYTE;
+    } else {
+        in_size = DEVICE_KEY_32BYTE;
+    }
+
+    trng_init(&trng_obj);
+
+    int ret = trng_get_bytes(&trng_obj, (unsigned char *)output, in_size, &size);
+    if (DEVICEKEY_SUCCESS != ret || in_size != size) {
```

Comment:
This is not valid use of the TRNG API -- `trng_get_bytes` is allowed to return less than what was requested, and in this case it's the responsibility of the caller to repeat the call until enough entropy has been gathered.

Reply:
@hanno-arm 
10x for that. Actually the documentation is not clear enough and I think missing an example. I have fixed it. Please review and let me know if this answer the API requirement.